GH-101357: Suppress `OSError` from `pathlib.Path.exists()` and `is_*()` (#118243)
python/cpython: Doc/library/pathlib.rst

.. method:: Path.is_dir(*, follow_symlinks=True)

>    Return ``True`` if the path points to an accessible directory, ``False``
>    otherwise. Use :meth:`Path.exists` to test if a path is accessible without
>    checking its type.

Reviewer: ```suggestion
   Return ``True`` if the path points to an accessible directory.
   Use :meth:`Path.exists` to determine if a path is not accessible,
   or the other ``is_*`` methods to test for other types.
```

How about this? Looking at it again, "False otherwise" seems redundant, and I think suggesting other `is_*` checks really emphasises that "`not is_dir()` doesn't imply `is_file()`".

Reviewer: Added a different suggestion further down (or up... depending on which way the review reads...)

.. method:: Path.is_dir(*, follow_symlinks=True)

   Return ``True`` if the path points to a directory. ``False`` will be

>    returned if the path is invalid, inaccessible or missing, or if it points
>    to something other than a directory. Use :meth:`Path.stat` to distinguish
>    between these cases.

Reviewer 2: Is it necessary to clarify this?

```suggestion
   returned if the path is invalid, inaccessible or missing, or if it points
   to something else. Use :meth:`Path.stat` to distinguish between
   these cases.
```

You could still make it shorter in that case:

```suggestion
   returned if the path is invalid, inaccessible or missing, or if it doesn't
   point to a directory. Use :meth:`Path.stat` to distinguish between these
   cases.
```

Author: https://github.com/python/cpython/pull/118243#discussion_r1600416339

.. method:: Path.is_symlink()

>    Return ``True`` if the path points to an accessible symbolic link,

Reviewer: Might be worth emphasising that this function does not follow symlinks? It's the odd one out, I believe.

Author: I added a bit about it returning `True` for broken links, which I think is a slightly more practical spin on the same idea. What do you think?

   ``False`` otherwise. Use :meth:`Path.exists` to test if a path is
   accessible without checking its type.

.. method:: Path.is_socket()

>    Return ``True`` if the path points to an accessible Unix socket, ``False``

Reviewer: Okay, I'm starting to regret the "accessible" at this point 😆 But I think I've figured out why:

"Accessible" is really a property of the _path_, but these descriptions make it sound like it's a property of the _file_ (or in this case, the socket). And "inaccessible" can be ambiguous for some of these (imagine a removable drive, where the _path_ may be accessible enough to determine the type, but the _drive_ is not accessible if nothing is in there).

So new proposal (for all of them):
> Return True if the path points to a Unix socket. False will be returned if the path is invalid, inaccessible or missing, or if it points to something other than a Unix socket. Use Path.exists to distinguish between these cases.

Reviewer 2: Isn't that a bit long? How about:

> Return True if the path is accessible and points to a valid Unix socket...

Reviewer: I'm really trying to get the reference to using Path.exists in there, to really clearly lead people towards that if they care about the different results (and thereby imply that "is_*" are for when you _don't_ care about the different results).

So while I do like the shorter phrasing, I think you then end up with all the same words describing the alternative ("Use Path.exists to distinguish between cases where the path is invalid, inaccessible, missing, or points to something other ..."). Given the choice, I'd rather more clearly explain _this_ function than the alternative (or not clearly explaining either of them).

Finding a balance between reference, specification, and teaching is the toughest part of documentation, and these docs have to do all three.

Reviewer 2: And this? If the path is accesible, it's also not missing.

> Return True if the path is accessible and points to a valid Unix socket. Use Path.exists to not check for the latter.

Reviewer: It's not just not checking the type, it's also going to raise different errors based on why it's inaccessible. But I didn't want to go as far as suggesting to handle `PermissionError` vs `FileNotFoundError` vs whatever other OSErrors may be raised. 

The main thing to imply is that a `True` result from these functions implies one case, while a `False` result implies one of many cases, to prevent people thinking that `not is_file()` is the same as `is_dir()` (which is obvious to us, who are neck deep in this stuff, but is not otherwise obvious at first sight). Suggesting that `exists` will let you distinguish exactly what `False` means is the best way I've come up with so far to communicate it without making it too hard to understand.

Also, I think "Use Path.exists to not check for the latter" is very difficult to parse. But even reworded for clarity, I don't think it would communicate what I'm wanting to communicate here.

Reviewer 2: You mean for Windows? Because for Unix the only difference is checking the file type:

https://github.com/python/cpython/blob/a7f83721d97f7b2b6c0efb21929b359c4143ed22/Lib/pathlib/_abc.py#L431-L435

https://github.com/python/cpython/blob/a7f83721d97f7b2b6c0efb21929b359c4143ed22/Lib/pathlib/_abc.py#L525-L528

Reviewer: Oh, does `exists` suppress all exceptions as well? If that's the case, we should be suggesting calling `stat` instead @barneygale

Reviewer 2: > Return True if the path points to a Unix socket. False is returned if the path is invalid, inaccessible, missing or points to something else. Use Path.stat to distinguish between these cases.

Author: Thanks for the feedback, both.

@nineteendo I did try the "something else" formulation, but being a bit more explicit is helpful for readers.
